crates/types/src/epoch_table.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+#![allow(dead_code)]
+
+use crate::identifiers::{LeaderEpoch, PartitionId};
+use crate::{GenerationalNodeId, Version, Versioned};
+use std::collections::HashMap;
+
+/// Partition processor epoch table.
+#[derive(Debug, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct EpochTable {
+    version: Version,
+    epochs: HashMap<PartitionId, EpochMetadata>,
+}
```

Comment:
Can we make the intention clear about how do we want to use this?
We have two things we want to store (potentially in separate places):
1. Last Epoch number for every partition ID (This can be in one table, or stored in metadata as a single key per partition ID). We can store observability data along with the value (+observability  info (which controller node decided, which node should be leader, timestamp of when was that epoch updated, reason for update) I'm assuming this is `EpochTable`
2. A state for cluster controller to record its partition distribution decisions, this is used to restore controller state and coordinate between controllers. This doesn't need to be synced with the rest of the nodes, but it also records "which node we think should run partitions x..y".

I'm not sure if this is also your intention or not.

Reply:
I've changed it so that we now store each epoch separately and the epoch is bumped not by the controller but by the worker.